plotly/figure_factory/_bullet.py

```diff
@@ -0,0 +1,257 @@
+                  measure_colors=None, subplot_spacing=None,
+                  title='Bullet Chart', height=600, width=1000):
+    """
+    Returns figure for bullet chart.
+
+    :param (pd.DataFrame | list) df: either a JSON list of dicts or a pandas
+        DataFrame. All keys must be one of 'title', 'subtitle', 'ranges',
+        'measures', and 'markers'.
+    :param (bool) as_rows: if True, the bars are placed horizontally as rows.
+        If False, the bars are placed vertically in the chart.
+    :param (int) marker_size: sets the size of the markers in the chart.
+    :param (str | int) marker_symbol: the symbol of the markers in the chart.
+        Default='diamond-tall'
+    :param (list) range_colors: a list of two colors between which all
+        the rectangles for the range are drawn. These rectangles are meant to
+        be qualitative indicators against which the marker and measure bars
+        are compared.
+        Default=['rgb(198, 198, 198)', 'rgb(248, 248, 248)']
+    :param (list) measure_colors: a list of two colors which is used to color
+        the thin quantitative bars in the bullet chart.
+        Default=['rgb(31, 119, 180)', 'rgb(176, 196, 221)']
+    :param (float) subplot_spacing: set the distance between each bar chart.
+        If not specified an automatic spacing is assigned based on the number
+        of bars to be plotted.
+    :param (str) title: title of the bullet chart.
+    :param (float) height: height of the chart.
+    :param (float) width width of the chart.
+    """
+    # validate df
+    if not pd:
+        raise exceptions.ImportError(
+            "'pandas' must be imported for this figure_factory."
+        )
+
+    if isinstance(df, list):
+        if not all(isinstance(item, dict) for item in df):
+            raise exceptions.PlotlyError(
+                'If your data is a list, all entries must be dictionaries.'
+            )
+        df = pd.DataFrame(df)
+
+    elif not isinstance(df, pd.DataFrame):
+        raise exceptions.PlotlyError(
+            'You must input a pandas DataFrame or a list of dictionaries.'
+        )
+
+    # check for valid keys
+    if any(key not in VALID_KEYS for key in df.columns):
+        raise exceptions.PlotlyError(
+            'Your headers/dict keys must be either {}'.format(
+                utils.list_of_options(VALID_KEYS, 'or')
+            )
+        )
+
+    # add necessary columns if missing
+    for key in VALID_KEYS:
+        if key not in df:
+            if key in ['title', 'subtitle']:
+                element = ''
+            else:
+                element = []
+            df[key] = [element for _ in range(len(df))]
+
+    # make sure ranges and measures are not NAN or NONE
+    for needed_key in ['ranges', 'measures']:
+        for idx, r in enumerate(df[needed_key]):
+            try:
+                r_is_nan = math.isnan(r)
+                if r_is_nan or r is None:
+                    df[needed_key][idx] = []
```

Comment:
similarly here. if you need to change data, make a copy of their data and then mutate the copy